### ISSUE TITLE
isVisibleForInteraction: fails to discover an overlay window

### ISSUE THREAD

**Issue by LeoNatan** [google/EarlGrey#735]
There seems to be an issue, where `isVisibleForInteraction:` returns `YES` despite there being a `UIWindow` instance that should take the touch of the user, rather than it reaching the element.

In addition to its current logic, the method should check first if there are other windows **visible and answering `hitTest:withEvent:` with a non-nil value**; if so, it should fail the interaction visibility check.

**Comment by tirodkar** [comment]
Thanks @LeoNatan. I'll investigate this and see if this would cause any issues. Visibility though is based on whether the element is visible on the screen and not whether the user can interact with it. Do you see this issue with grey_interactable as well?

**Comment by LeoNatan** [comment]
The issue stems from incorrect interaction. I just dug in the implementation and figured `isVisibleForInteraction:` is the culprit. 

The idea is, if there is an overlay window, any attempt of interaction with an element on a bottom window should fail if any window above the element swallows the touch event. This makes sense also from a user perspective (his touch is swallowed by the window). This can be the case even if, technically, the element is visible. It should just not be interactive.

**Comment by LeoNatan** [comment]
Another case is, if the element’s window or someone in the view hierarchy fails the hit test, interaction should also fail, even if visible.

**Comment by tirodkar** [comment]
I agree with your suggestion. We check if elements are interactable using grey_interactable() before the interaction. We can add the aforementioned check there. I'm not sure if making this change is correct for visibility since that only checks if the element is visible on the screen.

**Comment by LeoNatan** [comment]
Why are there different visibility checks? There is a regular visibility check, but then there is also visibility for interaction check.

But if there is a different method of checking interactivity, it makes sense to put it there.

**Comment by LeoNatan** [comment]
Ping. Any news on this issue?

**Comment by tirodkar** [comment]
I haven't looked into this in a bit. I'll send an update here once it happens. Is this a blocker for you atm?

**Comment by LeoNatan** [comment]
It's creating an issue in our Detox testing framework in specific types of scenarios where an overlay window is displayed.

**Comment by LeoNatan** [comment]
Thank you!
